File: 2022-scala/src/main/scala/scala-lib/Pos/Pos.scala
package scala-lib.Pos

final case class Pos(x: Int, y: Int) extends BoxPosOps[Pos] {
  def +(that: Pos): Pos =
    Pos(x + that.x, y + that.y)

  def *:(k: Int): Pos =
    Pos(k * x, k * y)

  def manhattanDistance(that: Pos): Int =
    (x - that.x).abs + (y - that.y).abs

  def <=(that: Pos): Boolean =
    x <= that.x && y <= that.y

  def min(that: Pos): Pos =
    Pos(x min that.x, y min that.y)

  def max(that: Pos): Pos =
    Pos(x max that.x, y max that.y)

  def getAxisOffsets: Seq[Pos] =
    Pos.axisOffsets.map(_ + Pos(x, y))

  def getDiagonalOffsets: Seq[Pos] =
    Pos.diagonalOffsets.map(_ + Pos(x, y))

  def getAllOffsets: Seq[Pos] =
    getAxisOffsets ++ getDiagonalOffsets

  def contains(that: Pos): Boolean =
    x <= that.x && that.y <= y

  def overlaps(that: Pos): Boolean =
    x <= that.y && that.x <= y
}

object Pos extends PosFactory[Pos] {
  val zero: Pos = Pos(0, 0)
  val axisOffsets: Seq[Pos] = Seq(Pos(0, 1), Pos(-1, 0), Pos(1, 0), Pos(0, -1))
  val diagonalOffsets: Seq[Pos] = Seq(Pos(-1, 1), Pos(1, 1), Pos(-1, -1), Pos(1, -1))
  val allOffsets: Seq[Pos] = axisOffsets ++ diagonalOffsets
}

File: csharp/csharp/csharp-lib/Pos/Pos.cs
namespace csharp.csharp_lib.Pos;

public record Pos(int X, int Y)
{
    public static Pos operator +(Pos p1, Pos p2) =>
        new(p1.X + p2.X, p1.Y + p2.Y);
    
    public static Pos operator -(Pos p1, Pos p2) =>
        new(p1.X - p2.X, p1.Y - p2.Y);

    public static Pos operator *(Pos p1, Pos p2) =>
        new(p1.X * p2.X, p1.Y * p2.Y);

    public static bool operator <=(Pos p1, Pos p2) =>
        p1.X <= p2.X && p1.Y <= p2.Y;

    public static bool operator >=(Pos p1, Pos p2) =>
        p1.X >= p2.X && p1.Y >= p2.Y;

    public bool Contains(Pos p2) =>
        X <= p2.X && Y <= p2.Y;

    public int ManhattanDistance(Pos that) =>
        Math.Abs(X - that.X) + Math.Abs(Y - that.Y);

    public IEnumerable<Pos> GetAxisOffsets() =>
        AxisOffsets.Select(pos => pos + new Pos(X, Y));

    public IEnumerable<Pos> GetDiagonalOffsets() =>
        DiagonalOffsets.Select(pos => pos + new Pos(X, Y));

    public IEnumerable<Pos> GetAllOffsets() =>
        GetAxisOffsets().Concat(GetDiagonalOffsets());
    
    public static Pos Zero = new(0, 0);

    private static readonly IEnumerable<Pos> AxisOffsets = new List<Pos>
    {
        new(-1, 0),
        new(0, -1),
        new(0, 1),
        new(1, 0),
    };

    private static readonly IEnumerable<Pos> DiagonalOffsets = new List<Pos>
    {
        new(-1, 1),
        new(1, 1),
        new(-1, -1),
        new(1, -1),
    };
}